common/src/main/java/com/gregtechceu/gtceu/data/recipe/builder/ShapedEnergyTransferRecipeBuilder.java

```diff
@@ -0,0 +1,205 @@
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.tags.TagKey;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.Ingredient;
+import net.minecraft.world.item.crafting.RecipeSerializer;
+import net.minecraft.world.level.ItemLike;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.Consumer;
+
+
+/**
+ * @author Irgendwer01
+ * @date 2023/11/4
+ * @implNote ShapedEnergyTransferRecipeBuilder
+ */
+public class ShapedEnergyTransferRecipeBuilder extends Builder<Ingredient, ShapedEnergyTransferRecipeBuilder> {
+    protected ItemStack output = ItemStack.EMPTY;
+    protected Ingredient chargeIngredient = Ingredient.EMPTY;
+    protected ResourceLocation id;
+    protected String group;
+    protected boolean transferMaxCharge;
+    protected boolean overrideCharge;
+
+    public ShapedEnergyTransferRecipeBuilder(@Nullable ResourceLocation id) {
+        this.id = id;
+    }
+
+    public ShapedEnergyTransferRecipeBuilder() {
+        this(null);
+    }
+
+    public ShapedEnergyTransferRecipeBuilder pattern(String slice) {
+        return aisle(slice);
+    }
+
+    public ShapedEnergyTransferRecipeBuilder define(char cha, TagKey<Item> itemStack) {
+        return where(cha, Ingredient.of(itemStack));
+    }
+
+    public ShapedEnergyTransferRecipeBuilder define(char cha, ItemStack itemStack) {
+        if (itemStack.hasTag() || itemStack.getDamageValue() >0) {
+            return where(cha, NBTIngredient.createNBTIngredient(itemStack));
+        }else {
+            return where(cha, Ingredient.of(itemStack));
+        }
+    }
+
+    public ShapedEnergyTransferRecipeBuilder define(char cha, ItemLike itemLike) {
+        return where(cha, Ingredient.of(itemLike));
+    }
+
+    public ShapedEnergyTransferRecipeBuilder define(char cha, Ingredient ingredient) {
+        return where(cha, ingredient);
+    }
+
+    public ShapedEnergyTransferRecipeBuilder chargeIngredient(Ingredient chargeIngredient) {
+        this.chargeIngredient = chargeIngredient;
+        return (ShapedEnergyTransferRecipeBuilder) this;
+    }
+
+    public ShapedEnergyTransferRecipeBuilder overrideCharge(boolean overrideCharge) {
+        this.overrideCharge = overrideCharge;
+        return (ShapedEnergyTransferRecipeBuilder) this;
+    }
+
+    public ShapedEnergyTransferRecipeBuilder transferMaxCharge(boolean transferMaxCharge) {
+        this.transferMaxCharge = transferMaxCharge;
+        return (ShapedEnergyTransferRecipeBuilder) this;
+    }
+
+    public ShapedEnergyTransferRecipeBuilder output(ItemStack itemStack) {
+        this.output = itemStack.copy();
+        return (ShapedEnergyTransferRecipeBuilder) this;
+    }
+
+    public ShapedEnergyTransferRecipeBuilder output(ItemStack itemStack, int count) {
+        this.output = itemStack.copy();
+        this.output.setCount(count);
+        return (ShapedEnergyTransferRecipeBuilder) this;
+    }
```

Comment:
The casts aren't needed here ^^

```suggestion
    public ShapedEnergyTransferRecipeBuilder chargeIngredient(Ingredient chargeIngredient) {
        this.chargeIngredient = chargeIngredient;
        return this;
    }

    public ShapedEnergyTransferRecipeBuilder overrideCharge(boolean overrideCharge) {
        this.overrideCharge = overrideCharge;
        return this;
    }

    public ShapedEnergyTransferRecipeBuilder transferMaxCharge(boolean transferMaxCharge) {
        this.transferMaxCharge = transferMaxCharge;
        return this;
    }

    public ShapedEnergyTransferRecipeBuilder output(ItemStack itemStack) {
        this.output = itemStack.copy();
        return this;
    }

    public ShapedEnergyTransferRecipeBuilder output(ItemStack itemStack, int count) {
        this.output = itemStack.copy();
        this.output.setCount(count);
        return this;
    }
```